presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java

```diff
@@ -0,0 +1,102 @@
+        extends CachingFileSystem
+{
+    private static final int BUFFER_SIZE = 65536;
+    private final boolean cacheValidationEnabled;
+    private boolean cacheQuotaEnabled;
+    private LocalCacheFileSystem localCacheFileSystem;
+
+    public AlluxioCachingFileSystem(ExtendedFileSystem dataTier, URI uri)
+    {
+        this(dataTier, uri, false);
+    }
+
+    public AlluxioCachingFileSystem(ExtendedFileSystem dataTier, URI uri, boolean cacheValidationEnabled)
+    {
+        super(dataTier, uri);
+        this.cacheValidationEnabled = cacheValidationEnabled;
+    }
+
+    @Override
+    public synchronized void initialize(URI uri, Configuration configuration)
+            throws IOException
+    {
+        this.localCacheFileSystem = new LocalCacheFileSystem(dataTier, uriStatus -> {
+            // CacheContext is the mechanism to pass the hiveFileContext to the source filesystem
+            // hiveFileContext is critical to use to open file.
+            CacheContext cacheContext = uriStatus.getCacheContext();
+            checkState(cacheContext instanceof PrestoCacheContext);
+            HiveFileContext hiveFileContext = ((PrestoCacheContext) cacheContext).getHiveFileContext();
+            try {
+                return dataTier.openFile(new Path(uriStatus.getPath()), hiveFileContext);
+            }
+            catch (Exception e) {
+                throw new IOException("Failed to open file", e);
+            }
+        });
+        this.cacheQuotaEnabled = configuration.getBoolean(USER_CLIENT_CACHE_QUOTA_ENABLED.getName(), false);
+        localCacheFileSystem.initialize(uri, configuration);
+    }
+
+    @Override
+    public FSDataInputStream openFile(Path path, HiveFileContext hiveFileContext)
+            throws Exception
+    {
+        // Using Alluxio caching requires knowing file size for now
+        if (hiveFileContext.isCacheable() && hiveFileContext.getFileSize().isPresent()) {
+            // FilePath is a unique identifier for a file, however it can be a long string
+            // hence using md5 hash of the file path as the identifier in the cache.
+            // We don't set fileId because fileId is Alluxio specific
+            FileInfo info = new FileInfo()
+                    .setLastModificationTimeMs(hiveFileContext.getModificationTime())
+                    .setPath(path.toString())
+                    .setFolder(false)
+                    .setLength(hiveFileContext.getFileSize().get());
+            String cacheIdentifier = md5().hashString(path.toString(), UTF_8).toString();
+            // CacheContext is the mechanism to pass the cache related context to the source filesystem
+            CacheContext cacheContext = PrestoCacheContext.build(cacheIdentifier, hiveFileContext, cacheQuotaEnabled);
+            URIStatus uriStatus = new URIStatus(info, cacheContext);
+            FSDataInputStream cachingInputStream = localCacheFileSystem.open(uriStatus, BUFFER_SIZE);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource: *cachingInputStream*. The resource is referenced at line 95. The resource is returned at line 98. There are other execution paths that don&#39;t close the resource or return it, for example, when ExtendedFileSystem.openFile throws an exception, and an if condition at line 94 evaluates to true. To prevent this resource leak, close *cachingInputStream* along these other paths before you exit this method.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).